crates/sqlexec/src/remote/staged_stream.rs

```diff
@@ -0,0 +1,161 @@
+use futures::{Future, FutureExt};
+use std::collections::HashMap;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use crate::errors::{internal, Result};
+use parking_lot::Mutex;
+use tokio::sync::oneshot;
+use uuid::Uuid;
+
+use super::exchange_stream::ClientExchangeRecvStream;
+
+pub type StagedClientStreams = StagedStreams<ClientExchangeRecvStream>;
+
+/// Hold streams that pending for execution.
+pub struct StagedStreams<S> {
+    /// Streams keyed by broadcast id.
+    streams: Mutex<HashMap<Uuid, PendingStream<S>>>,
+}
+
+/// State of the pending stream.
+///
+/// Since we need to coordinate between client and server, we have two states:
+/// - Client begins streaming first.
+/// - Server begins executing first.
+///
+/// Both states are correct since we have no "scheduler" triggering which
+/// happens first. And so while the client will always send the plan to the
+/// server first, the server may actually begin executing that plan prior to the
+/// client actually starting to stream.
+enum PendingStream<S> {
+    StreamArrivedFirst(S),
+    WaitingForStream(oneshot::Sender<S>),
+}
+
+impl<S> StagedStreams<S> {
+    /// Put a stream for later execution.
+    pub fn put_stream(&self, id: Uuid, stream: S) {
+        let mut streams = self.streams.lock();
+
+        // Handle case where we began executing before the stream arrived.
+        if let Some(pending) = streams.remove(&id) {
+            match pending {
+                PendingStream::StreamArrivedFirst(_) => panic!("attempted to put stream twice"), // Programmer bug.
+                PendingStream::WaitingForStream(channel) => {
+                    // We don't care if the receiver dropped. Means it was
+                    // canceled on the "execution" side.
+                    let _ = channel.send(stream);
+                    return;
+                }
+            }
+        }
+
+        // Handle case where stream arrived first.
+        streams.insert(id, PendingStream::StreamArrivedFirst(stream));
+    }
+
+    /// Resolve a pending stream by id.
+    ///
+    /// This will wait until we have the stream available.
+    pub fn resolve_pending_stream(&self, id: Uuid) -> ResolveStreamFut<S> {
```

Comment:
wouldn't we want to do some housekeeping after the stream is resolved & clear the entry from `StagedStreams`?

Reply:
Yep we're removing from the map here.

We will need to add in something to remove stale streams (e.g. server starts executing, inserts into the map, then client errors causing the entry to never be removed).